src/server/string_family.h

```diff
@@ -16,6 +17,28 @@ class CommandRegistry;
 using facade::OpResult;
 using facade::OpStatus;
 
+// Stores a string, the pending result of a tiered read or nothing
+struct StringValue {
+  StringValue() : v_{} {
+  }
+  StringValue(std::string s) : v_{std::move(s)} {
+  }
+  StringValue(util::fb2::Future<std::string> f) : v_{std::move(f)} {
+  }
+
+  // Get and consume value. If backed by a future, blocks until resolved.
+  std::string Get() &&;
+
+  // If no value is stored
+  bool IsEmpty() const;
+
+  // Read string from prime value - either from memory or issue tiered storage read
+  static StringValue Read(std::string_view key, const PrimeValue& pv, EngineShard* es);
+
+ private:
+  std::variant<std::monostate, std::string, util::fb2::Future<std::string>> v_;
```

Comment:
I still think it's more elegant to move this functionality into Future - afaik, sophisticated future implementations support fast-paths as well.

Reply:
food for thought

Reply:
I saw the comment in helio. Let's merge this and I will digest the food for thought 😆